src/shogun/mathematics/graph/Array.h

```diff
@@ -0,0 +1,241 @@
+#include <shogun/mathematics/graph/nodes/Input.h>
+#include <shogun/mathematics/graph/nodes/MatMul.h>
+
+namespace shogun
+{
+	namespace graph
+	{
+		class LazyExpr
+		{
+
+		protected:
+			std::unordered_map<
+			    std::shared_ptr<node::Input>, std::shared_ptr<Tensor>>
+			    m_inputs;
+			std::shared_ptr<node::Node> m_output;
+			struct Protected
+			{
+			};
+
+		public:
+			LazyExpr(const std::shared_ptr<Tensor>& input_tensor)
+			{
+				auto input = std::make_shared<node::Input>(
+				    input_tensor->get_shape(),
+				    input_tensor->get_type()->type());
+				m_inputs.emplace(input, input_tensor);
+				m_output = input;
+			}
+
+			LazyExpr(const std::unordered_map<
+			    std::shared_ptr<node::Input>, std::shared_ptr<Tensor>>& inputs, 
+			    std::shared_ptr<node::Node>& output, Protected): m_inputs(inputs), m_output(output)
+			{
+			}
+
+			LazyExpr(Protected)
+			{
+			}
+
+			template <typename T>
+			void bind_input(const T& expr)
+			{
+				if constexpr (std::is_same_v<T, std::unique_ptr<LazyExpr>>)
+				{
+					for (const auto& [input_node, input_tensor] :
+					     expr->get_inputs())
+					{
+						if (!m_inputs.count(input_node))
+						{
+							m_inputs.emplace(input_node, input_tensor);
+						}
+					}
+				}
+			}
+
+			/** bind rhs nodes to lhs (this)
+			 */
+			template <
+			    typename OperatorType, typename... Args,
+			    std::enable_if_t<std::is_base_of_v<node::Node, OperatorType>>* =
+			        nullptr>
+			void bind(Args&&... args)
+			{
+				(bind_input(std::forward<Args>(args)), ...);
+
+				m_output = std::make_shared<OperatorType>(
+				    m_output, return_node(std::forward<Args>(args))...);
+			}
+
+			const std::unordered_map<
+			    std::shared_ptr<node::Input>, std::shared_ptr<Tensor>>&
+			get_inputs() const
+			{
+				return m_inputs;
+			}
+
+			const std::shared_ptr<node::Node> get_output() const
+			{
+				return m_output;
+			}
+
+			[[nodiscard]] std::unique_ptr<LazyExpr> copy()
+			{
+				return std::make_unique<LazyExpr>(m_inputs, m_output, Protected{});
+			}
+
+		private:
+			template <typename T>
+			auto return_node(const T& expr)
+			{
+				if constexpr (std::is_same_v<T, std::unique_ptr<LazyExpr>>)
+				{
+					return expr->get_output();
+				}
+				else
+				{
+					return expr;
+				}
+			}
+		};
+
+		/**
+		 *
+		 *	auto A = std::make_shared<Array>(mat);
+		 *	auto B = std::make_shared<Array>(mat);
+		 *
+		 *	auto C = A + B;
+		 *	decltype(C) == decltype(A) == decltype(B);
+		 */
+		class Array
+		{
+		protected:
+			struct Protected
+			{
+			};
+
+			std::unique_ptr<LazyExpr> m_lazy_expr;
+			std::shared_ptr<Tensor> m_output_tensor;
+
+		public:
+			Array(Protected)
+			{
+			}
+
+			template <typename T>
+			Array(const T& input) : m_output_tensor(std::make_shared<Tensor>(input))
+			{
+				m_lazy_expr = std::make_unique<LazyExpr>(m_output_tensor);
+			}
+
+			template <typename T, std::enable_if_t<std::is_rvalue_reference_v<T&&>>* = nullptr>
```

Comment:
@vigsterkr does this look correct to you? Without the SFINAE `T&&` just matches everything